R/raincloudPlots.R

```diff
@@ -0,0 +1,908 @@
+  vioPosVec <- c()
+  for (i in vioNudgeForEachCloud) {
+    vioPosVec <- c(vioPosVec, rep(i, 512))  # Each violin consists of 512 points by default
+  }
+  vioArgsPos <- list(
+    width = options$vioHeight, position = ggplot2::position_nudge(x = vioPosVec), side = vioSides
+  )
+
+  # Box positioning
+  boxArgsPos <- list(width = options$boxWidth, position = boxPosition)
+
+  # Point positioning
+  negativePointNudge <- if (!options$customSides) {
+    options$pointNudge * -1  # This way all nudges in the GUI are positive by default
+  } else {
+    0                        # CustomSides fixes points to Axis ticks (see HelpButton in .qml)
+  }
+  jitter <- if (!options$jitter) 0 else NULL
+  pointArgsPos   <- list(
+    position = ggpp::position_jitternudge(
+      nudge.from = "jittered",
+      width      = options$pointSpread, # xJitter
+      x          = negativePointNudge,  # Nudge
+      height     = jitter,              # Jitter, particularly interesting for likert data
+      seed       = 1                    # Reproducible jitter
+    )
+  )
+
+  # Cov and id
+  covArg <- if (options$covariate == "")                                 NULL else "covariate"  # Must be string
+  idArg  <- if (options$observationId   == "" || options$primaryFactor == "")  NULL else "observationId"
+            # primaryFactor condition necessary because if user input primaryFactor, then adds observationId input,
+            # and then removes primaryFactor again, JASP/GUI/qml will not remove observationId input
+
+  # Call geom_rain()
+  output <- ggrain::geom_rain(
+
+    violin.args = vioArgs, boxplot.args = boxArgs, point.args = pointArgs, line.args = lineArgs,
+
+    rain.side        = NULL,  # Necessary for neat positioning
+    violin.args.pos  = vioArgsPos,
+    boxplot.args.pos = boxArgsPos,
+    point.args.pos   = pointArgsPos,
+    line.args.pos    = pointArgsPos,  # Dependent on points
+
+    cov         = covArg,
+    id.long.var = idArg,
+    likert      = FALSE  # TRUE won´t work because of ggpp:position_jitternudge() in pointArgsPos
+                         # instead use height argument in ggpp:position_jitternudge()
+  )  # End geom_rain()
+
+  return(output)
+}  # End .rainGeomRain()
+
+
+
+# .rainOutlineColor() ----
+# Sets the Outline color of violins and boxes
+.rainOutlineColor <- function(options, inputOutline, infoFactorCombinations) {
+
+  output <- NULL
+
+  if (inputOutline == "black") {
+    output <- rep("black", infoFactorCombinations$numberOfClouds)
+  } else if (inputOutline == "none") {
+    output <- rep(NA, infoFactorCombinations$numberOfClouds)
+  } else if (inputOutline == "colorPalette") {
+    if (options$secondaryFactor != "" || options$colorAnyway) {
+      output <- infoFactorCombinations$colors
+    } else {
+      output <- rep("black", infoFactorCombinations$numberOfClouds)
+    }
+  } else {
+    print("error with Outline")
```

Comment:
should this be an error the user sees?

Reply:
no, that was a left-over from debugging. i removed the else statement altogether.